fix: Parenting does not preserve WorldPositionStays option and other parenting related issues

NoelStephensUnity opened the pull request:
When parenting and setting the WorldPositionStays to false on the server, clients would not apply the same WorldPositionStays setting because it was not being synchronized.  This PR also provides parent child serialization ordering during the initial client synchronization that happens when the connection is approved.
[MTT-4616](https://jira.unity3d.com/browse/MTT-4616)

### For Visual Validation via Manual Testing:
Use the companion test project PR #2207

## Changelog
- Added: Position, rotation, and scale to the `ParentSyncMessage` which provides users the ability to specify the final values on the server-side when `OnNetworkObjectParentChanged` is invoked just before the message is created (when the `Transform` values are applied to the message).
- Added: `NetworkObject.TryRemoveParent` method for convenience purposes opposed to having to cast null to either `GameObject` or `NetworkObject`.

- Fixed: Issue where the `WorldPositionStays` parenting parameter was not being synchronized with clients. 
- Fixed: Issue where parented in-scene placed `NetworkObject`s would fail for late joining clients. 
- Fixed: Issue where scale was not being synchronized which caused issues with nested parenting and scale when `WorldPositionStays` was true. 
- Fixed: Issue with `NetworkTransform.ApplyTransformToNetworkStateWithInfo` where it was not honoring axis sync settings when `NetworkTransformState.IsTeleportingNextFrame` was true.
- Fixed: issue with `NetworkTransform.TryCommitTransformToServer` where it was not honoring the `InLocalSpace` setting.

## Testing and Documentation
- Includes integration test that validates with and without `WorldPositionStays`.
- Includes integration test for in-scene placed multi-generation parented `NetworkObjects`.
- Includes integration test for in-scene placed `NetworkObject` nested under a GameObject with no NetworkObject
- Documentation updates are necessary. ([PR-778](https://github.com/Unity-Technologies/com.unity.multiplayer.docs/pull/778))

NoelStephensUnity committed:
fix

This fixes the issue with the client-side keeping the world position when parenting during initial spawning and when parenting post-spawn.

NoelStephensUnity committed:
fix

We should set the position to zero and the quaternion to identity when parenting with the world position stays disabled.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
style

updated comment

NoelStephensUnity committed:
style

minor addition to the same comment.

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -583,10 +584,11 @@ internal void SetCachedParent(Transform parentTransform)
         internal (bool IsReparented, ulong? LatestParent) GetNetworkParenting() => (m_IsReparented, m_LatestParent);
 
-        internal void SetNetworkParenting(bool isReparented, ulong? latestParent)
+        internal void SetNetworkParenting(bool isReparented, ulong? latestParent, bool childWorldPositionStays)
```
Is `childWorldPositionStays` the right name ?
I’d have expected `myWorldPositionStays` or `worldPositionStays`.

JesseOlmer replied:
Agreed - let's not rename the already-existing Unity parameter name to a new name.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -997,12 +1001,14 @@ public struct TransformData : INetworkSerializeByMemcpy
                 writeSize += Header.HasParent ? FastBufferWriter.GetWriteSize(ParentObjectId) : 0;
                 writeSize += Header.HasTransform ? FastBufferWriter.GetWriteSize(Transform) : 0;
-                writeSize += Header.IsReparented ? FastBufferWriter.GetWriteSize(IsLatestParentSet) + (IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0) : 0;
+                writeSize += Header.IsReparented ? FastBufferWriter.GetWriteSize(IsLatestParentSet) + FastBufferWriter.GetWriteSize<bool>() + (IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0) : 0;
```
```suggestion
                if (Header.IsReparented)
                {
                    writeSize += FastBufferWriter.GetWriteSize(IsLatestParentSet);
                    writeSize += FastBufferWriter.GetWriteSize(ChildWorldPositionStays);
                    writeSize += IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0;
                }
```

This line was already borderline unreadable and now it's even worse.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1053,7 +1060,7 @@ public unsafe void Deserialize(FastBufferReader reader)
                 var readSize = Header.HasParent ? FastBufferWriter.GetWriteSize(ParentObjectId) : 0;
                 readSize += Header.HasTransform ? FastBufferWriter.GetWriteSize(Transform) : 0;
-                readSize += Header.IsReparented ? FastBufferWriter.GetWriteSize(IsLatestParentSet) + (IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0) : 0;
+                readSize += Header.IsReparented ? FastBufferWriter.GetWriteSize(IsLatestParentSet) + FastBufferWriter.GetWriteSize<bool>() + (IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0) : 0;
```
```suggestion
                if (Header.IsReparented)
                {
                    readSize += FastBufferWriter.GetWriteSize(IsLatestParentSet);
                    readSize += FastBufferWriter.GetWriteSize(ChildWorldPositionStays);
                    readSize += IsLatestParentSet ? FastBufferWriter.GetWriteSize<ulong>() : 0;
                }
```

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1074,6 +1081,7 @@ public unsafe void Deserialize(FastBufferReader reader)
                 if (Header.IsReparented)
                 {
                     reader.ReadValue(out IsLatestParentSet);
+                    reader.ReadValue(out ChildWorldPositionStays);
```
nit: Feels wrong injecting this between "IsLatestParentSet" and "latestParent" - those seem a pair so should be next to each other... same on the write side of things, obviously.

NoelStephensUnity committed:
Update-JOlmer-Suggestions

Code clarity.

Co-authored-by: Jesse Olmer <jesseo@unity3d.com>

NoelStephensUnity committed:
Update-JOlmer-Suggestion-2

Code clarity.

Co-authored-by: Jesse Olmer <jesseo@unity3d.com>

NoelStephensUnity committed:
update

Changing the property name to m_WorldPositionStays.
Re-organized the order in which we write and read this value.
Only writing the value in the ParentSyncMessage if we are reparented (not sure why we wouldn't be).

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

ShadauxCat commented:
Code looks good... will approve once tests are committed.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
fix

Setting WorldPositionStays when sending  ParentSyncMessage in OnTransformParentChanged.
To better handle late joining players with synchronizing NetworkObjects, added parent child sorting to the SceneEventData's synchronization serialization process.

Added world position stays parameter to CreateLocalNetworkObject to preserve worldPostionStays.

NoelStephensUnity committed:
test

Adding the integration test for this fix.

NoelStephensUnity committed:
update

Setting position before parenting.
Added the same logic for in-scene placed NetworkObjects.

NoelStephensUnity committed:
test update

Testing with and without WorldPositionStays.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
update

removing unnecessary conditions in the  SortParentedNetworkObjects metod.
updated the comments to reflect the changes.

NoelStephensUnity committed:
fix

This fixes some edge case scenarios and issues with parenting in-scene placed NetworkObjects.
One in particular is to provide the local space position and rotation in ParentSyncMessage if the WorldPositionStays value is false.  The server might maintain the original local space values and upon being reparented it re-applies the local space values.  This is only pertinent if the user decides they don't want to use a NetworkTransform for the NetworkObject in question.

Fixed an issue with despawning objects if the WorldPositionStaysTest fails during the integration test. Migrating this to teardown resolves the issue.

NoelStephensUnity committed:
style

removing extra CR/LF

NoelStephensUnity committed:
fix

fixing reversed order of local vs world space rotation assignment.

NoelStephensUnity committed:
test

Updating the test with varying child position and rotation values.

NoelStephensUnity committed:
update

adding two entries to the changelog.

NoelStephensUnity committed:
update

adjusted one of the entries for clarity.

NoelStephensUnity committed:
style

Updating comment to further explain the need for the added local space position and rotation properties.

NoelStephensUnity committed:
fix

Fixing missed size calculation adjustment.
Made deserialization read size calculations look like the serialization write size calculations for code readability/clarity.

NoelStephensUnity committed:
update

Reverting minor adjustment to read size calculation.

NoelStephensUnity committed:
refactor

CreateLocalNetworkObject now just accepts the NetworkObject.SceneObject structure as opposed to the 10 properties of the NetworkObject.SceneObject structure passed in as parameters.
Minor code clean up when assigning position and rotation.

NoelStephensUnity committed:
fix

This fixes several issues with parenting.
Now, a users can control what the final position, rotation, and scale of a NetworkObject is when being added/removed to/from a parent's sibling list ("parented or de-parented").
This also makes it possible to invoke NetworkObject.TrySetParent with a null value in order to be able to remove a parent while also selecting whether they want WorldPositionStay to be true or false (when removing a parent WorldPositionStays impacts the final transform values and can lead to scale issue if you parent with WorldPositionStays set to false and then it remove the parent with WorldPositionStays set to true).

NoelStephensUnity committed:
style

adjusting comments based on recent updates

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -576,6 +576,7 @@ internal void InvokeBehaviourOnNetworkObjectParentChanged(NetworkObject parentNe
         private bool m_IsReparented; // Did initial parent (came from the scene hierarchy) change at runtime?
         private ulong? m_LatestParent; // What is our last set parent NetworkObject's ID?
         private Transform m_CachedParent; // What is our last set parent Transform reference?
+        private bool m_WorldPositionStays = true; // Used to preserve the world position stays parameter passed in TrySetParent
```
Maybe `m_CachedWorldPositionStays` to signify it's not really an "active" setting?

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -598,7 +600,10 @@ internal void SetNetworkParenting(bool isReparented, ulong? latestParent)
+            var networkObject = parent.GetComponent<NetworkObject>();
+
+            // If the parent doesn't have a NetworkObjet then return false, otherwise continue trying to parent
+            return networkObject == null ? false : TrySetParent(networkObject, worldPositionStays);
```
Why is behavior different here vs `TrySetParent(GO)` and `TrySetParent(NO)`?
Pass in something without a networkobject here and you get False and no-op.
Pass in something without a networkobject on `TrySetParent(GO)` and you get a call to `TrySetParent(NO)`
Pass in something without a networkobject to `TrySetParent(NO)` and you turn the thing into a root object.

NoelStephensUnity replied:
You cannot pass a null value in since this overloaded version of TrySetParent is a Transform (struct).
So, the appropriate response here if the transform's `GameObject` does not have a `NetworkObject` component is to return false.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -640,17 +654,13 @@ public bool TrySetParent(NetworkObject parent, bool worldPositionStays = true)
                 return false;
             }
 
-            if (parent == null)
```
It looks like you used to be unable to un-parent on this API, but now you've introduced that capability here... was that intentional? If so, is it a different issue/feature from the WorldPositionStays bug?

NoelStephensUnity replied:
Well, the issue was that we were not synchronizing the WorldPositionStays parameter.  In order to fix this on both sides (parenting and de-parenting) we need to provide the same flow since `transform.SetParent` should use the same WorldPositionStays value (particularly when it is false).

Before a user would have to either use transform.parent = null (which defaults to WorldPositionStays == true) or use transform.SetParent (which bypasses some of our netcode specific checks).

This allows a user to use NetworkObject.TrySetParent for both parenting and de-parenting.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -640,17 +654,13 @@ public bool TrySetParent(NetworkObject parent, bool worldPositionStays = true)
+            m_WorldPositionStays = worldPositionStays;
 
-            transform.SetParent(parent.transform, worldPositionStays);
+            transform.SetParent(parent ? parent.transform : null, worldPositionStays);
```
I'm not sure if this is safe, or if it bypasses the usual marshall to native null-checking Unity does.

NoelStephensUnity replied:
Made changes to avoid that potential issue.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -712,17 +722,33 @@ private void OnTransformParentChanged()
+            // We always set this true to be able to apply the parenting
+            // If it is being de-parented, then m_IsReparented will be set
+            // to false.
+            // TODO: Determine if m_IsReparented is still needed.
```
Fix TODO or remove it?

NoelStephensUnity replied:
Will look further into this.

NoelStephensUnity replied:
So... this was my conundrum:
m_IsReparented was never getting set to false once it was set (server-side) and so once something was parented then it would always be true.  So the two "semi-hidden" places to look at would be `NetworkObject.GetMessageSceneObject` and `NetworkSpawnManager.OnDespawnObject` that both use the `NetworkObject.GetNetworkParenting`.

Both uses of "isReparented" depend upon the latestParent having a value.  Which really, we could remove m_IsReparented and just determine if it has a parent or not (without the reparented) as both cases are just trying to determine if it needs to set a parent or not.

Let me make some adjustments and get rid of the TODOs

NoelStephensUnity replied:
I went ahead and removed the m_IsReparented

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -712,17 +722,33 @@ private void OnTransformParentChanged()
             }
 
+            // We always set this true to be able to apply the parenting
+            // If it is being de-parented, then m_IsReparented will be set
```
I guess the answer to this depends on the TODO below, but unparenting (turning into a root object) could still mean it's been "reparented" (that is, it no longer has the parent it did at authoring time), so why set to false?

NoelStephensUnity replied:
ApplyNetworkParenting basically exits if reparented is false.
It gets set to true before it is invoked and if it is de-parenting then, server-side, it resets it to false.
Let me look at the last place it is actually being used to see if we can just remove it... of course we would need to change the documentation a bit.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -712,17 +722,33 @@ private void OnTransformParentChanged()
             var message = new ParentSyncMessage
             {
                 NetworkObjectId = NetworkObjectId,
-                IsReparented = m_IsReparented,
```
Was it intentional to remove this parameter? Seems premature to remove it if the TODO above is still around...

NoelStephensUnity replied:
this no longer is used.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -712,17 +722,33 @@ private void OnTransformParentChanged()
-                LatestParent = m_LatestParent
+                LatestParent = m_LatestParent,
+                WorldPositionStays = m_WorldPositionStays,
+                Position = m_WorldPositionStays ? transform.position : transform.localPosition,
+                Rotation = m_WorldPositionStays ? transform.rotation : transform.localRotation,
+                Scale = transform.localScale,
```
Does this play nice when there IS a NetworkTransform on the object, or do things start fighting?

NoelStephensUnity replied:
It plays nice with NetworkTransform.  
This is only set once when the ParentSyncMessage is sent to connected clients.
A NetworkTransform would basically update the same values.
Cost of bandwidth here does increase but that is dependent upon when the ParentSyncMessage is sent relative to the end of the current network tick when the deltas are sent by the NetworkTransform (they very well could be different if both messages are sent on different frames).

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -762,6 +788,7 @@ internal bool ApplyNetworkParenting()
                 return false;
             }
 
+            // TODO: Determine if m_IsReparented is still needed.
```
Fix or remove, please

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -770,22 +797,30 @@ internal bool ApplyNetworkParenting()
             if (m_LatestParent == null || !m_LatestParent.HasValue)
             {
                 m_CachedParent = null;
-                transform.parent = null;
-
+                // We must use Transform.SetParent when taking WorldPositionStays into
+                // consideration, otherwise just setting transform.parent = null defaults
+                // to WorldPositionStays which can cause scaling issues if the parent's
+                // scale is not the default (Vetctor3.one) value.
+                transform.SetParent(null, m_WorldPositionStays);
+                // TODO: Determine if m_IsReparented is still needed.
+                m_IsReparented = false;
                 InvokeBehaviourOnNetworkObjectParentChanged(null);
                 return true;
             }
 
             if (!NetworkManager.SpawnManager.SpawnedObjects.ContainsKey(m_LatestParent.Value))
             {
-                OrphanChildren.Add(this);
+                if (!OrphanChildren.Contains(this))
+                {
+                    OrphanChildren.Add(this);
+                }
```
Why did you make this change? It adds an additional lookup and doesn't change the functionality at all... from reference-source:
```
        void ICollection<T>.Add(T item) {
            AddIfNotPresent(item);
        }
```
and from the docs about the boolean return value:
```
true if the element is added to the [HashSet<T>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-6.0) object; false if the element is already present.
```

NoelStephensUnity replied:
That was a mistake and am removing it.
👍

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -998,12 +1033,24 @@ public struct TransformData : INetworkSerializeByMemcpy
 
             public int NetworkSceneHandle;
 
+            public bool WorldPositionStays;
```
Public field, needs docs. Also, do we need to add this? If so, you need to change this from `fix` to `feat`

NoelStephensUnity replied:
This is a public field of an internal struct NetworkObject.SceneObject.
Does that rule apply to public fields of internally defined structs?

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -998,12 +1033,24 @@ public struct TransformData : INetworkSerializeByMemcpy
                 writeSize += Header.HasParent ? FastBufferWriter.GetWriteSize(ParentObjectId) : 0;
+                if (Header.HasParent)
+                {
+                    writeSize += FastBufferWriter.GetWriteSize(ParentObjectId);
```
Are you double-counting ParentObjectId here (and on line 1041 above)?

NoelStephensUnity replied:
Doh >.<
That other like should be deleted.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -314,40 +314,52 @@ internal bool HasPrefab(NetworkObject.SceneObject sceneObject)
+            if (sceneObject.Header.HasParent)
             {
-                if (SpawnedObjects.TryGetValue(parentNetworkId.Value, out NetworkObject networkObject))
+                if (SpawnedObjects.TryGetValue(sceneObject.ParentObjectId, out NetworkObject networkObject))
```
Why are you reading from sceneObject after you just created `parentNetworkId`?

NoelStephensUnity replied:
Yeah I missed that... good catch...

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -314,40 +314,52 @@ internal bool HasPrefab(NetworkObject.SceneObject sceneObject)
+                else if (isReparented) // TODO: Not sure we need this warning any longer
                 {
-                    if (NetworkLog.CurrentLogLevel <= LogLevel.Normal)
+                    if (NetworkLog.CurrentLogLevel <= LogLevel.Developer)
```
Seems bad - is this a warning, a developer info, an internal SDK bug?

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -314,40 +314,52 @@ internal bool HasPrefab(NetworkObject.SceneObject sceneObject)
             }
 
-            if (!NetworkManager.NetworkConfig.EnableSceneManagement || !isSceneObject)
+            if (!NetworkManager.NetworkConfig.EnableSceneManagement || !sceneObject.Header.IsSceneObject)
```
`sceneObject.Header.IsSceneObject`... should the variable be called "sceneObject" if it's not a sceneobject? :)

NoelStephensUnity replied:
Yeah the name actually I think we could change...this seems to be what it was called for sometime now.
I seem to remember creating like a NetworkObject.Serialize and NetworkObject.Deserailize a long time ago and this turned into "SceneObject" from the message itself.

NoelStephensUnity replied:
@JesseOlmer 
This was the only area that I felt could be another PR since this one has started to become relatively large.
Would you like me to change the name of NetworkObject.SceneObject in this PR? 
If so, do you have any suggestions for the name change?

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -243,7 +243,25 @@ internal void AddSpawnedNetworkObjects()
+            // as well as the order in which they will be deserialized
+            if (m_NetworkManager.LogLevel == LogLevel.Developer)
+            {
+                var message = "[Server-Side Client-Synchronization] NetworkObject serialization order:\n";
```
nit: StringBuilder for multiple appends like this.

NoelStephensUnity replied:
It is done! (good suggestion)

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -323,6 +341,32 @@ private int SortNetworkObjects(NetworkObject first, NetworkObject second)
             return 0;
         }
 
+        /// <summary>
```
This feels like scope creep. Don't we already have deferred parenting support? And doesn't this not correctly handle sorting a hierarchy nesting level > 1?

NoelStephensUnity replied:
If you enable developer logging and run the associated integration test you will see that it nests 10 children (each child is a parent and has a child) and is spawned in reverse order to make children spawned (server-side) before their respective parent.  When clients are connecting and synchronizing, the final synchronization serialization order yields the parents before the children (i.e. it sorts a hierarchy nesting level > 1).

It only sorts during synchronization to help assure that a child will always have a parent available.
When dynamically spawning and parenting, then the orphan child list is used.

NoelStephensUnity replied:
Sorting them for client synchronization helps when there are multi-generation nested children that all have varying scales. It helps assure that it applies from the root parent down to the last nested child in regards to the initial client synchronization.
_I was thinking this would be a nice place to provide user implementations to provide users the ability to write their own sorting orders for the initial client synchronization based on any unique needs they might have._

ShadauxCat replied:
Could this be handled using DeferredMessageManager?

If a spawn is received for an object that has a parent set, but the parent doesn't exist, CreateObjectMessage can do this:

```csharp

if (ObjectInfo.HasParent && !networkManager.SpawnManager.SpawnedObjects.ContainsKey(ObjectInfo.ParentObjectId))
{
    networkManager.DeferredMessageManager.DeferMessage(IDeferredMessageManager.TriggerType.OnSpawn, ObjectInfo.ParentObjectId, reader, ref context);
    return false;
}
```

I think that'd handle this case and also other potential edge cases with less complexity.

ShadauxCat replied:
(The way that works, if it's not obvious, is that on the receive side, if it receives the messages in the wrong order, it defers them with a specific trigger. The OnSpawn trigger means that, when an object with a NetworkObjectId matching the second parameter is spawned, the deferred message is processed immediately... so basically it just does this same sorting operation for you, it just does it on the receiving end in a generic way.)

NoelStephensUnity replied:
This is only for initial client synchronization.  So there is one really large message that has all NetworkObjects in it.  We can remove this, but it does remove the chance of a child being added to the orphaned children list during the initial synchronization and if user code depends upon a parent existing when a child is spawned the parent is guaranteed to have already been spawned and parented by the time OnNetworkSpawn is invoked on the child.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -323,6 +341,32 @@ private int SortNetworkObjects(NetworkObject first, NetworkObject second)
+        private int SortParentedNetworkObjects(NetworkObject first, NetworkObject second)
+        {
+            // If the first has a parent, move the first down
+            if (first.transform.parent != null)
```
if they BOTH have parents, return 0? This logic will currently return 1.

NoelStephensUnity replied:
The sorting occurs on the first object.
If the first object has a parent, then it is a child and move it down.
Otherwise...if the first does not have a parent but the second does have a parent then move the first up (i.e. children move down the list).
If neither the first or the second have a parent then don't move the first element.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -314,40 +314,52 @@ internal bool HasPrefab(NetworkObject.SceneObject sceneObject)
                     if (parentNetworkObject != null)
                     {
-                        networkObject.transform.SetParent(parentNetworkObject.transform, true);
+                        networkObject.TrySetParent(parentNetworkObject.transform, worldPositionStays);
```
What's the rationale for not setting position & rotation if there's a handler?

NoelStephensUnity replied:
Let me review this area.

NoelStephensUnity replied:
So, the position and rotation are sent to the prefab handler and the prefab handler determines whether it will apply those values or not (up to the user).

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -387,14 +399,31 @@ internal NetworkObject CreateLocalNetworkObject(bool isSceneObject, uint globalO
                     networkObject.NetworkManagerOwner = NetworkManager;
 
+                    // Go ahead and set network parenting properties
+                    networkObject.SetNetworkParenting(isReparented, parentNetworkId, worldPositionStays);
+
+                    if (sceneObject.Header.HasTransform)
+                    {
+                        if (worldPositionStays)
+                        {
+                            networkObject.transform.position = position;
+                            networkObject.transform.rotation = rotation;
+                        }
+                        else
+                        {
+                            networkObject.transform.localPosition = position;
+                            networkObject.transform.localRotation = rotation;
+                        }
+                    }
+
                     if (parentNetworkObject != null)
                     {
-                        networkObject.transform.SetParent(parentNetworkObject.transform, true);
+                        networkObject.TrySetParent(parentNetworkObject.transform, worldPositionStays);
                     }
```
nit: you're doing this 2 times, maybe 3 with prefabhandler above... feels like a good candidate for a private method?

NoelStephensUnity replied:
Are you talking about lines 402 - 427 and lines 439 - 472?

NoelStephensUnity replied:
@JesseOlmer 
I consolidated this code.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Spawning/NetworkSpawnManager.cs:
```diff
@@ -419,9 +448,27 @@ internal NetworkObject CreateLocalNetworkObject(bool isSceneObject, uint globalO
+                }
+
+                // Go ahead and set network parenting properties
+                networkObject.SetNetworkParenting(isReparented, parentNetworkId, worldPositionStays);
```
Does the order matter here? You're doing it differently than above...

NoelStephensUnity replied:
Ahh yeah I could move that.  Really, both `SetNetworkParenting` and setting the position and rotation need to occur before `NetworkObject.TrySetParent`.  Whether the position and rotation is set before `SetNetworkParenting`  or vice versa doesn't matter as much as they should be set prior to trying to set the parent.
However, I think keeping the same order of operations is cleaner and easier to understand.

NoelStephensUnity replied:
This was resolved when I consolidated this section of code.

JesseOlmer commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ParentSyncMessage.cs:
```diff
@@ -1,29 +1,43 @@
+                BytePacker.WriteValuePacked(writer, (ulong)LatestParent);
             }
+
+            BytePacker.WriteValuePacked(writer, Position);
+            BytePacker.WriteValuePacked(writer, Rotation);
+            BytePacker.WriteValuePacked(writer, Scale);
```
Are we actually getting any space savings on these? They're a bunch of unbounded floats and bytepacker just unceremoniously casts the memory location to a UInt8 before trying to pack it. I'd expect this to actually bloat the size of the message since now we're probably serializing an entire Uint8 PLUS a byte for the "packed size" for each of these...

NoelStephensUnity committed:
update

Applying some of the suggested changes.
Fixed issue where ParentSyncMessage was calculating too large of a buffer size.

NoelStephensUnity committed:
update

Removing packing for values where there are no savings.

NoelStephensUnity committed:
update

Renaming m_WorldPositionStays to m_CachedWorldPositionStays

NoelStephensUnity committed:
update

Using the local parentNetworkId as opposed to the sceneObject's.
Keeping the same order of operations for the setting of position and rotation and setting the network parenting values.

NoelStephensUnity committed:
update

Removed m_IsReparented and associated properties.
Also consolidated the CreateLocalNetworkObject's assignment of the transform values, parenting, and DDOL migration.

NoelStephensUnity committed:
style

Updating comments with m_CachedWorldPositionStays in place of m_WorldPositionStays

NoelStephensUnity committed:
update

preserve non-networkobject parenting.
WIP:
It might be easiest to just adjust ParentSyncMessage to have a single bool that signifies the NetworkObject should be de-parented.

NoelStephensUnity committed:
update

Finalizing the removal of the m_IsReparented property.
Added a RemoveParent to the ParentSync message.
Added additional comments

NoelStephensUnity committed:
fix

Parenting and Scale:
Fixed issue where GetMessageSceneObject wasn't using the NetworkObject's lossy scale value.  This would result in improper scaling of parented (WorldPositionStays = true) child NetworkObjects for late joining clients.  Added comments with details on this.
Added applying scale to CreateLocalNetworkObject.

Further refined the parenting process by:
Removed parenting from CreateLocalNetworkObject
Removed setting the cached parent directly within SpawnNetworkObjectLocallyCommon as this is done within NetworkObject.ApplyNetworkParenting.

NetworkTransform:
Fixed issue where TryCommitTransformToServer was not honoring the NetworkTransform.InLocalSpace property.
Fixed issue where teleporting did not honor the NetworkTransforms's current synchronize axis settings.

NoelStephensUnity committed:
test updates

ParentingTestsGeneral (still WIP):
Now validates scale with multi-generation nested children where most of the parents have a scale other than Vector3.one. This validates dynamically spawned NetworkObjects (currently) only. Player NetworkObjects scale is validated in existing NetworkTransformTests.

NetworkTransformStateTest:
As I was verifying everything I realized this test was testing way too many redundant combinations and it wasn't fully testing everything for the NetworkTransformState.
I also updated the bool parameters to be more meaningful named enums so you know what exactly is being tested.
Added additional NetworkTansformState.IsTeleportingNextFrame logic to the existing test.
Added an additional "step" which verifies the NetworkTransformState is additively collapsing for each detected axial delta over the relative threshold value.

NoelStephensUnity committed:
update

Implementing Jesse's suggested use of StringBuilder for better performance and not continually create a new string for each NetworkObject name added to the final message.

Added a comment in ParentSyncMessage regarding why we apply transform values after parenting.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkTransform/NetworkTransformStateTests.cs:
```diff
@@ -4,15 +4,87 @@
+
         {
+            bool inLocalSpace = m_TransformSpace == TransformSpace.Local;
+            bool isTeleporting = synchronizationType == SynchronizationType.Teleport;
```
Don't hate me for the below code... :)
This was the best way I could come up with that would be sure to cover all of the possible combinations to assure each axial sync bool value was set when it needed to be set.
We could have more detailed Assert messages, but for now we know specifically what set of synchronized axis failed, whether it was when teleporting or just a delta update, and what space (world or local) it occurred within by the enum  parameters used.

NoelStephensUnity committed:
style

cleaning up some code that I had modified while debugging.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -495,14 +495,18 @@ protected void TryCommitTransformToServer(Transform transformToCommit, double di
             }
             else // Non-Authority
             {
+                var position = InLocalSpace ? transformToCommit.localPosition : transformToCommit.position;
```
Noticed this while working on parenting with multi-generation nested children (there was a bug elsewhere it turned out).  I went ahead and fixed it as I have more parenting integration tests I am working on that will be using this method when the parented NetworkObjects have a NetworkTransform too.

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -609,63 +598,63 @@ private bool ApplyTransformToNetworkStateWithInfo(ref NetworkTransformState netw
             }
 
-            if (SyncPositionX && Mathf.Abs(networkState.PositionX - position.x) >= PositionThreshold || networkState.IsTeleportingNextFrame)
+            if (SyncPositionX && (Mathf.Abs(networkState.PositionX - position.x) >= PositionThreshold || networkState.IsTeleportingNextFrame))
```
This bug I found while debugging parent-child scale issues when the NetworkObject also had a NetworkTransform.  I had turned the NetworkTransform scale synchronization flags (X, Y,Z) off but it seemed to keep updating them...which made me realize the initial state after 1 or 2 network ticks is a "teleport state"...which the previous logic (without the braces) would apply all axis whether marked to be synchronized or not.

NoelStephensUnity committed:
test update

Turning off verbose debug.
Cleaning the code up a bit.
Only waiting for scale condition when using a NetworkTransform.

NoelStephensUnity committed:
update

updating changelog to reflect the recent updates.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
update

Removing the private m_TickFrequency property as it is no longer used.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
test update

Renamed the ParentingTestsGeneral to ParentingWorldPositionStaysTest.
Created abstract IntegrationTestWithApproximation to avoid replicating the exact same set of Approximately methods.
Minor style clean up (spelling issues and extra CR/LFs)

NoelStephensUnity committed:
fix

This fixes some issues with the very first initialization of multigeneration child nested in-scene placed NetworkObjects.

NoelStephensUnity committed:
style

Updating comments and some formating.

NoelStephensUnity committed:
test

These are assets that will be used for an integration test I am currently working on.
This is still WIP.

NoelStephensUnity committed:
style

missed a white space >.<

NoelStephensUnity committed:
fix

This fixes one edge case scenario where nested in-scene placed NetworkObject children are de-parented and then a client late joins. Under this scenario (only for in-scene placed NetworkObjects) we want the late joining client to remove its parent, then set the transform values, and then get spawned.

NoelStephensUnity committed:
test in-scene placed

This includes all of the scripts and assets that the ParentingInSceneObjectsTests needs.
It includes adding the ParentingInSceneObjects scene to the scenes in build list as well.

NoelStephensUnity committed:
update

Migrating the added scene for integration testing to the bottom of the scenes in build list.

NoelStephensUnity committed:
update and style

Since the next update will be a minor revision, I added the TryRemoveParent helper method to make removing a parent easier.  Without this method users will have to use TrySetParent and cast the null value to either GameObject or NetworkObject.

Updated some of the comments for clarity.

NoelStephensUnity committed:
test

the last test for this PR that verifies an in-scene placed NetworkObject parented under a GameObject will preserve that hierarchy.

NoelStephensUnity committed:
update

Adding one more entry for this PR

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

jeffreyrainy commented on com.unity.netcode.gameobjects/Components/NetworkTransform.cs:
```diff
@@ -1007,7 +983,6 @@ public override void OnNetworkSpawn()
         {
             m_CachedIsServer = IsServer;
             m_CachedNetworkManager = NetworkManager;
-            m_TickFrequency = 1.0 / NetworkManager.NetworkConfig.TickRate;
```
💯 nice unused variable

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ParentSyncMessage.cs:
```diff
@@ -1,29 +1,49 @@
+        public bool RemoveParent;
+
+        // These additional properties are used to synchronize clients with the current position
+        // , rotation, and scale after parenting/de-parenting (world/local space relative). This
```
What's the leading " , " before rotation? Can we send the comma one line up, right after "position" ?

NoelStephensUnity replied:
yeah, that does look funny... it is the 1 whitespace requirement after the start of a comment line that makes it look odd.
It should be:  position, rotation, and scale
But...when you put that comma on the next line it does seem visually perplexing now that you brought it to my attention.
I can move the comma up to the previous line.

NoelStephensUnity committed:
style

adjusting a comma to the previous comment line for readability.

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
style

adjusted lossyScale comment that was missing "if".

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

NoelStephensUnity committed:
Merge branch 'develop' into fix/parenting-does-not-preserve-worldpositionstays-option

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -749,42 +790,90 @@ private void OnTransformParentChanged()
-                return true;
+                var parentNetworkObject = transform.parent.GetComponent<NetworkObject>();
+
+                // If the parent is a GameObject then preserve that hierarchy
```
Nit... Comment says "if the parent is a GameObject", but the check is actually checking if it's a NetworkObject

NoelStephensUnity replied:
Ahh... yeah I need to improve the clarity here... 
If the parent is just a GameObject (without a NetworkObject...which would mean parentNetworkObject is null)... but it could be rephrased to be less confusing indeed.

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1051,36 +1144,48 @@ public unsafe void Deserialize(FastBufferReader reader)
+                    // We need to read at this point in order to get the IsLatestParentSet value
+                    if (!reader.TryBeginRead(readSize))
+                    {
+                        throw new OverflowException("Could not deserialize SceneObject: Out of buffer space.");
```
Nit: I'd make this error something like "Trying to read past the end of the buffer" rather than "out of buffer space". "Out of buffer space" could be interpreted as an Out Of Memory issue (i.e., couldn't allocate a big enough buffer to read it).

NoelStephensUnity replied:
Sure, I can adjust that language.
_(I actually just used the same message if trying to read the header failed.)_

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1051,36 +1144,48 @@ public unsafe void Deserialize(FastBufferReader reader)
                 }
 
-                if (Header.HasParent)
+                if (IsLatestParentSet)
```
Not entirely related to this change, but it's just occurred to me that we could make these ulongs into longs and effectively use the sign bit to encode whether they're set or not... read one byte, if it's negative it's unset, if it's positive then seek back one byte and read it as ulong. That is, assuming the actual value of this doesn't ever need that bit...

Could be a useful function in BytePacker... WriteOptionalValue() that writes 8 bytes with a 0 leading bit if set, or 1 byte with a 1 leading bit if unset.

Curious what you'd think of that (not an action item for this PR, just a thought)

ShadauxCat replied:
Actually it also occurs to me that there are several boolean values here...

What about writing them as a bitmask using `writer.EnterBitwiseContext()`?

NoelStephensUnity replied:
I think we should have a separate PR where we go through all `INetworkMessage` implementations and see if we can reduce the total size for each message type.  Perhaps not in this PR?

ShadauxCat commented on com.unity.netcode.gameobjects/Runtime/Messaging/Messages/ParentSyncMessage.cs:
```diff
@@ -1,29 +1,49 @@
                 if (IsLatestParentSet)
                 {
-                    writer.WriteValueSafe((ulong)LatestParent);
+                    BytePacker.WriteValuePacked(writer, (ulong)LatestParent);
```
ooh, fancy.

...but might I suggest looking at `BytePacker.WriteValueBitPacked()` and see if it's viable? It may not be. It has better compression than `WriteValuePacked()` but for a ulong it does limit the allowed value to only 61 bits, as it uses the first 3 bits to encode a byte count. (As opposed to `WriteValuePacked()`, which uses a whole byte to encode the byte count on top of the actual value itself, and if all bytes of the value are needed, it actually makes the encoded value larger than the original instead of smaller)

ShadauxCat replied:
(In contrast, `WriteValueBitPacked()` guarantees it will never increase the size of the value, but does have a limitation of only 61 bits... also there's a typo in the comment that says it's limited to 31 bits, but it's not, it's 61.)

NoelStephensUnity committed:
style

Updating comment for clarity as per Kitty's suggestion.

NoelStephensUnity committed:
update

Replacing WriteValuePacked and ReadValuePacked with WriteValueBitPacked and ReadValueBitPacked.

netcode-ci-service committed:
Merge develop into fix/parenting-does-not-preserve-worldpositionstays-option

netcode-ci-service committed:
Merge develop into fix/parenting-does-not-preserve-worldpositionstays-option
